client/src/hooks/useSearch.js

```diff
@@ -0,0 +1,85 @@
+import { useRouter } from 'next/router'
+
+export const useSearch = (state, setState) => {
+  const search = state
+  const setSearch = setState
+  const router = useRouter()
+
+  const goToSearchResults = (goToStart = false) => {
+    if (goToStart) delete search.query.offset
+    const { pathname, query } = search
+    router.push({ pathname, query })
+  }
+
+  const setSearchString = (searchString, resetQuery = false) => {
+    // should this clear the search facets?
+    if (resetQuery) search.query = {}
+    if (searchString === '') {
+      delete search.query.search
+    } else {
+      search.query.search = searchString
+    }
+
+    setSearch({ ...search })
+  }
+
+  // get a set of unique items we can add to or delete from
+  const getSetFromFacet = (facet) => {
+    const existingFacet = search.query[facet]
+    if (!existingFacet) return new Set()
+    if (Array.isArray(existingFacet)) return new Set(existingFacet)
+    return new Set([existingFacet])
+  }
+
+  const addFacet = (facet, value) => {
+    if (value) {
+      const newFacet = getSetFromFacet(facet)
+      newFacet.add(value)
+      search.query[facet] = [...newFacet]
```

Comment:
Is it not possible to just update the search object to a set for query instead of a list everywhere rather than converting back and forth?

Reply:
it can't be a set unless we transform it to an array before sending the query to the next router. Also when next supplies the query from the route it doesn't know if some query parameters should be arrays or a string:

`/search?category=DATASET`
would have category as `{category: 'DATASET'}`
and 
`/search?category=DATASET&category=MODEL_ORGANISM`
would have category as `{category: ['DATASET', 'MODEL_ORGANISM']}`
also 
`/search?category=DATASET`
can be created by `{category: ['DATASET']}` or {category: 'DATASET'}
but it will always be the latter when supplied from next

Using set just makes this easier but it does need to be switch back to an array at some point, I figured this was nicer as opposed to accessing a query with a getter method instead

Reply:
Ah word. Yeah this is a nice way to do this. Just wanted to make sure we couldn't just use a set everywhere instead.